libdrgn/kdump_reader.c

```diff
@@ -0,0 +1,286 @@
+					 key, kdump_get_err(ctx));
+	}
+
+	switch (attr.type) {
+	case KDUMP_STRING:
+		*offset = strtoull(attr.val.string, NULL, 16);
+		break;
+	case KDUMP_NUMBER:
+		*offset = attr.val.number;
+		break;
+	default:
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "%s - unexpected attr type: %d\n",
+					 key, attr.type);
+	}
+
+	kdump_attr_unref(ctx, &root);
+	return NULL;
+}
+
+struct drgn_error *drgn_kdump_get_osrelease(kdump_ctx_t *ctx,
+                                            char *osrelease)
+{
+	const char *key = "linux.uts.release";
+
+	kdump_attr_ref_t root;
+	kdump_status ks = kdump_attr_ref(ctx, key, &root);
+	if (ks != KDUMP_OK) {
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	kdump_attr_t attr;
+	ks = kdump_attr_ref_get(ctx, &root, &attr);
+	if (ks != KDUMP_OK) {
+		kdump_attr_unref(ctx, &root);
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref_get(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	switch (attr.type) {
+	case KDUMP_STRING:
+		strncpy(osrelease, attr.val.string, OSRELEASE_LEN);
+		break;
+	default:
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "%s - unexpected attr type: %d\n",
+					 key, attr.type);
+	}
+
+	kdump_attr_unref(ctx, &root);
+	return NULL;
+}
+
+struct drgn_error *drgn_kdump_get_page_size(kdump_ctx_t *ctx,
+                                            uint64_t *page_size)
+{
+	const char *key = KDUMP_ATTR_PAGE_SIZE;
+
+	kdump_attr_ref_t root;
+	kdump_status ks = kdump_attr_ref(ctx, key, &root);
+	if (ks != KDUMP_OK) {
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	kdump_attr_t attr;
+	ks = kdump_attr_ref_get(ctx, &root, &attr);
+	if (ks != KDUMP_OK) {
+		kdump_attr_unref(ctx, &root);
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "kdump_attr_ref_get(%s) failed: %s\n",
+					 key, kdump_get_err(ctx));
+	}
+
+	switch (attr.type) {
+	case KDUMP_NUMBER:
+		*page_size = attr.val.number;
+		break;
+	default:
+		return drgn_error_format(DRGN_ERROR_OTHER,
+		                         "%s - unexpected attr type: %d\n",
+					 key, attr.type);
+	}
+
+	kdump_attr_unref(ctx, &root);
+	return NULL;
+}
+
+static bool drgn_kdump_is_64bits(const char *kdump_arch_attr)
+{
+	return (!strcmp(kdump_arch_attr, KDUMP_ARCH_X86_64) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_AARCH64) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_ALPHA) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_IA64) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_PPC64) ||
+	        !strcmp(kdump_arch_attr, KDUMP_ARCH_S390X));
+}
+
+struct drgn_error *drgn_kdump_get_arch(kdump_ctx_t *ctx,
+                                       enum drgn_architecture_flags *arch)
```

Comment:
My stack trace work is going to conflict with this bit since I'm changing around how drgn represents the architecture, but I can work that out on my end.

Reply:
I'll keep an eye too in case your change makes it first.